cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java

```diff
@@ -0,0 +1,575 @@
+          + "gidhycxS86dxpEljnOMCw8CKoUBd5I880IUahEiUltk7OLJYS/Ts1wbn3kPOVX3wyJs8WBDtBkFrDHW2ezth2QJ"
+          + "ADj3e1YhMVdjJW5jqwlD/VNddGjgzyunmiZg0uOXsHXbytYmsA545S8KRQFaJKFXYYFo2kOjqOiC1T2cAzMDjCQ"
+          + "==\n-----END PRIVATE KEY-----\n";
+
+  private MockStsTransportFactory mockStsTransportFactory;
+  private static MockTokenServerTransportFactory mockTokenServerTransportFactory;
+
+  static class MockStsTransportFactory implements HttpTransportFactory {
+    MockStsTransport transport = new MockStsTransport();
+
+    @Override
+    public HttpTransport create() {
+      return transport;
+    }
+  }
+
+  @Before
+  public void setUp() {
+    mockStsTransportFactory = new MockStsTransportFactory();
+    mockStsTransportFactory.transport.setReturnAccessBoundarySessionKey(true);
+
+    mockTokenServerTransportFactory = new MockTokenServerTransportFactory();
+    mockTokenServerTransportFactory.transport.addServiceAccount(
+        "service-account@google.com", "accessToken");
+  }
+
+  @Test
+  public void fetchIntermediateCredentials() throws Exception {
+    GoogleCredentials sourceCredentials =
+        getServiceAccountSourceCredentials(mockTokenServerTransportFactory);
+
+    ClientSideCredentialAccessBoundaryFactory factory =
+        ClientSideCredentialAccessBoundaryFactory.newBuilder()
+            .setSourceCredential(sourceCredentials)
+            .setHttpTransportFactory(mockStsTransportFactory)
+            .build();
+
+    IntermediateCredentials intermediateCredentials = factory.fetchIntermediateCredentials();
+
+    // Verify requested token type.
+    Map<String, String> query =
+        TestUtils.parseQuery(mockStsTransportFactory.transport.getRequest().getContentAsString());
+    assertEquals(
+        OAuth2Utils.TOKEN_TYPE_ACCESS_BOUNDARY_INTERMEDIARY_TOKEN,
+        query.get("requested_token_type"));
+
+    // Verify intermediate token and session key.
+    assertEquals(
+        mockStsTransportFactory.transport.getAccessBoundarySessionKey(),
+        intermediateCredentials.getAccessBoundarySessionKey());
+    assertEquals(
+        mockStsTransportFactory.transport.getAccessToken(),
+        intermediateCredentials.getIntermediateAccessToken().getTokenValue());
+  }
+
+  @Test
+  public void fetchIntermediateCredentials_withCustomUniverseDomain() throws IOException {
+    String universeDomain = "foobar";
+    GoogleCredentials sourceCredentials =
+        getServiceAccountSourceCredentials(mockTokenServerTransportFactory)
+            .toBuilder()
+            .setUniverseDomain(universeDomain)
+            .build();
+
+    ClientSideCredentialAccessBoundaryFactory factory =
+        ClientSideCredentialAccessBoundaryFactory.newBuilder()
+            .setUniverseDomain(universeDomain)
+            .setSourceCredential(sourceCredentials)
+            .setHttpTransportFactory(mockStsTransportFactory)
+            .build();
+
+    factory.fetchIntermediateCredentials();
```

Comment:
Can this call return an exception since `foobar` is a non-valid universe domain? I think we can keep the verify domain test case below to ensure that the url is constructed properly

Reply:
`ClientSideCredentialAccessBoundaryFactory` doesn't have a way of checking if a domain is valid or not. The only thing that the builder checks is that the domain on factory matches the source credential and we have a test that covers that scenario (`public void builder_universeDomainMismatch_throws()`) Since we're using a mock STS transport factory, it doesn't throw an exception here.

Reply:
I see. It looks like they're failing at two different spots (one is in the builder and one is in the refresh logic).

If possible, can the `GoogleCredentials` object above be mocked? Mock the `GoogleCredentials.refresh()` call so that it will throw an exception (i.e. url with the custom endpoint turns out to be invalid)?

Reply:
Oh wait, I see this scenario is covered in `refreshCredentialsIfRequired_sourceCredentialCannotRefresh_throwsIOException`.